backend: Support sending stream messages by stream ID.

eeshangarg opened the pull request:
This is a follow-up to #10391.

@timabbott FYI :)

showell commented:
Can you add some explanation to the final commit (in the commit comments)?  It feels like a lot is happening there.

timabbott commented on zerver/lib/addressee.py:
```diff
@@ -111,17 +117,22 @@ def legacy_build(sender: UserProfile,
                 if sender.default_sending_stream:
                     # Use the users default stream
-                    stream_name = sender.default_sending_stream.name
+                    stream = sender.default_sending_stream.name
```
Can you add a commit on top of this making this code path not go back through stream name, given that we're starting with a stream object?

eeshangarg replied:
Sorry, but I am not sure what you mean. We kinda have to use the stream name here because the caller relies on the name to call `get_stream`. Also, `stream` isn't a `Stream` object, it can be either an ID or a stream name, in this case (maybe `stream_name_or_id` might be a better variable name, I'll change it to that). Or am I missing something here? :)

timabbott replied:
`sender.default_sending_stream` is a Stream object.  So if we now support in the following code having a stream ID or a stream name, we could do `.id` here, not `.name`.

timabbott commented:
@eeshangarg thanks for working on this!  I merged a couple of the prep commits.

timabbott commented on zerver/lib/addressee.py:
```diff
@@ -85,9 +85,8 @@ def user_profiles(self) -> List[UserProfile]:
-    def stream_name(self) -> str:
+    def stream_name(self) -> Optional[str]:
         assert(self.is_stream())
-        assert(self._stream_name is not None)
```
What's the story with this?

eeshangarg replied:
Actually, I am not sure what the purpose of this assertion is. By the time someone calls `stream_name()`, the stream should already have been validated. Especially, now that an `Addressee` may have an ID and no name, this assertion is useless because we need a way to determine whether an ID or name was provided and that implies that `stream_name()` can return `None`.

We could also just choose to have a generic `stream()` method that returns a `Stream` object, and handles the work of choosing between the ID or name internally.

timabbott replied:
I assume it was to make sure we ahve valid input.

eeshangarg commented:
@timabbott: Just updated this and responded to your feedback/questions. I have one question, what should we do about functions like `internal_send_stream_message`, should we be changing those to support IDs?

@showell: Thanks for the suggestion, I just updated the commit description for the last commit with more details about some of the auxiliary changes. While I could have moved these to separate commits (especially the change to `send_pm_if_empty_stream`), note that simply including them in the commit that also changes `check_message` makes everything testable using only one test and avoids brittle tests that only test helper methods. At least, that's how I approached it! Please let me know if there is a better way to structure changes like this! Thanks!

timabbott commented:
> @timabbott: Just updated this and responded to your feedback/questions. I have one question, what should we do about functions like internal_send_stream_message, should we be changing those to support IDs?

We'll want to think about this -- what do the callers of `internal_send_stream_message` generally have?  If it's a stream object, then we should at least do the stream ID (if not the Stream object).

eeshangarg commented:
@timabbott: I just updated this. So, I did some digging, and here's what I found:

* The only place (apart from `check_message`) where `Addressee.stream_name()` is called is in `_internal_prep_message`, which calls `ensure_stream` to create a new stream with the given stream name if need be. Since `Addressee.stream_name()` can now return `None`, we have to make sure that `ensure_stream` is only called if `stream_name()` *doesn't* return `None`. So, I pushed a commit that amends that conditional and added an explanatory comment.

* Now, `_internal_prep_message` is called by `internal_prep_stream_message`, which is called by `internal_send_stream_message`. As for `internal_send_stream_message`, its callers do not have access to the `Stream` object, so we need not touch it!

* As for `internal_prep_stream_message`, it is called in less than 5 places and in some of those, we do have access to the `Stream` object. However, I noticed that our subscription API relies on `internal_prep_stream_message`. And I could be wrong, but I remember that someone calling the API can subscribe to a non-existent stream and the stream will be created if it doesn't exist, so the subscription API depends on stream names (and consequently the eventual call to `ensure_stream`) anyway. So, I decided not to touch this function either because the other cases (`prep_stream_welcome_message` and `zilencer/management/commands/add_mock_conversation.py`) didn't seem important enough to warrant changing this function to support stream IDs. Having said that, please let me know if you'd like me to revisit this for reasons I may be unaware of!

Keeping the above points in mind, once the above commits can be merged, as far as I know, that should conclude the backend portion of this project! Looking forward to hearing your feedback! Thanks! :)

timabbott commented:
@eeshangarg merged, after rewriting the comment in the last commit to give less convoluted reasoning.  Fundamentally, what's actually going on here is that if you have a stream ID, the stream already exists.

I do think we should still do a follow-up project to make the `internal_send_stream` flows support passing a Stream object (rather than a name), since we're currently doing an unnecessary database query with unnecessary complexity.  But that probably deserves it's own issue; it's certainly less important than this API change.

I guess documentation and API bindings are next?

timabbott commented:
Thanks for doing this, by the way; it was a pretty involved refactor.
